### ISSUE TITLE
GL textures display checkerboard placeholders of varying duration

### ISSUE THREAD

**Issue by oldschooljarvis** [Famous/engine#272]
I initially reproduced this issue on Slack as follows:

https://gist.github.com/oldschooljarvis/2011f7d9ea0586c1f059
Just `````famous init````` a new project and replace `````src/index.js````` with the above ^

![bugreproduction4-1_01](https://cloud.githubusercontent.com/assets/3048285/8120328/7e32e6b8-1063-11e5-9920-2ac8cbf95199.gif)

The sequence of events is:

`````(0ms)````` First node is spawned containing a DOMElement utilizing the default logo image.
`````(4000ms)````` Second node is spawned with a Mesh instead, using the same default logo image but mapped as a GL texture. There is visible checkerboarding on load.
`````(8000ms)````` Third node is spawned, identical to the second, but does not checkerboard. Ditto all subsequent nodes.

The checkerboarding only manifests itself if it's a fresh page load. It will not manifest upon refreshing the page.

-

https://gist.github.com/oldschooljarvis/031463a0f8cb8fbf3d84
http://i.imgur.com/be3bldF.png/famous_logo_magenta.png (second image)

![bugreproduction4-2_01](https://cloud.githubusercontent.com/assets/3048285/8120506/6bb44232-1065-11e5-8216-b9f42fd1f5ca.gif)

Similar setup, except we eschew the DOM node and deal purely in GL.

Sequence of events is:

``````(1000ms)`````` Spawn node w/ default logo. - Bad checkerboxing, lasts nearly a second.
``````(2000ms)`````` Spawn node w/ default logo again. - Zero checkerboarding.
``````(5000ms)`````` Spawn node w/ modified logo image. - Slight checkerboarding, maybe a couple of frames. **Note: The .gif fails to capture this at all.**
`````(6000ms)````` Spawn node w/ modified logo image. - Zero checkerboarding.

However, if we refresh the page, checkerboarding is eliminated entirely:

![bugreproduction4-2_02](https://cloud.githubusercontent.com/assets/3048285/8120518/952d5310-1065-11e5-8e11-52a9ab0b96a9.gif)

-

The first reproduction demonstrated that the issue isn't related to normal browser image caching.

The second reproduction demonstrated that loading a new GL texture incurs only minor checkerboarding; that subsequent nodes of the same texture don't have the issue, and that the first texture in a scene is always afflicted with the worst checkerboarding. Perhaps the latter behavior is something to do with initialization of the texture system itself? Just a thought.

These issues were reproduced on Chrome Version 43.0.2357.124 m (64-bit), Windows 8.1. The system I tested on is a 3.8GHz i7 w/ an R9 270. Firefox and IE both exhibit the issue just the same.

**Comment by FarhadG** [comment]
Thanks for informative report. I believe @adnan-wahab will be looking into this issue very soon.

**Comment by adnan-wahab** [comment]
Hey @oldschooljarvis, the checkerboard is a placeholder that indicates that the texture has not been fully loaded. The image is first loaded across the network, and then onto the gpu. In this code, the network request is only happening once, but uploading the texture to the gpu is happening too many times. Ideally, images should only be registered in the texture registry once. Check out this container for an example  https://api-te.famo.us/codemanager/v1/containers/0e578483-207f-44ae-8ac8-3762962dfa31/share

**Comment by oldschooljarvis** [comment]
>In this code, the network request is only happening once, but uploading the texture to the gpu is happening too many times. Ideally, images should only be registered in the texture registry once.

In my production code I actually properly fetch the texture from ```TextureRegistry``` with subsequent instances. Forgot to do that here. While I agree it's certainly not optimal, I don't believe it's the cause of any demonstrated latent loading/checkerboarding. In the tests above, you can see there were zero issues with subsequent texture loads once a texture was already registered, even if it was being registered twice.

Regardless, in the second example I provided, registration and subsequent loading of an entirely new texture takes a dramatically shorter amount of time provided that it's not the first texture to be registered in the scene.

Basically what I'm looking for is a way to know if a texture has been loaded and is ready for use. Perhaps an event or callback.

>Check out this container for an example https://api-te.famo.us/codemanager/v1/containers/0e578483-207f-44ae-8ac8-3762962dfa31/share

I'm getting a blank screen there, not sure if it should just work or I need to embed it somehow (haven't used containers or the deploy system at all yet).

-

Thanks for your help.

**Comment by oldschooljarvis** [comment]
For what it's worth, ```TextureManager``` maintains its own internal texture registry, and for each texture registered, there is a ```isLoaded``` property. This is also where the temporary checkerboard placeholder is assigned. Unfortunately, ```TextureRegistry``` doesn't have any such property.

I could try to tell if a texture is loaded by checking the value of ```isLoaded``` with something like:
```f.core.FamousEngine.getContext(x).getWebGLRenderer().textureManager.registry[i].isLoaded```,
but I'm not sure that's the best way to go about it.

**Comment by michaelobriena** [comment]
@adnan-wahab can you look into this?
